### PR TITLE
Add support for `additional_contexts` in `build` service config

### PR DESCRIPTION
**Pull request by laurazard**
**What I did**

Implement support for `additional_contexts` as per https://github.com/compose-spec/compose-spec/pull/307 and https://github.com/compose-spec/compose-go/pull/354.

**Related issue**

closes https://github.com/docker/compose/issues/9461
closes https://github.com/docker/compose/issues/9961

**(not mandatory) A picture of a cute animal, if possible in relation to what you did**

image

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -2,6 +2,8 @@ module github.com/docker/compose/v2
 
 go 1.20
 
+replace github.com/compose-spec/compose-go => ../../../compose-spec/compose-go
```
Comment (laurazard) [inline go.mod:5]: just a reminder here so we don't merge this until a new compose-go release + removing this replace

File: pkg/compose/build.go [hunk 2]
```diff
@@ -402,6 +405,14 @@ func getImageBuildLabels(project *types.Project, service types.ServiceConfig) ty
 	return ret
 }
 
+func parseBuildContexts(additionalContexts map[string]*string) map[string]build.NamedContext {
```
Comment (ndeloof) [inline pkg/compose/build.go:408]: nit: AFAICT we don't `parse` things here, so better name this `toBuildContexts`

**Review by ndeloof — APPROVED** [review]

**Comment by laurazard** [comment]
@glours PTAL :)

**Review by ndeloof — APPROVED** [review]

**Comment by ndeloof** [comment]
looks like we don't have coverage report for cucumber tests

**Comment by laurazard** [comment]
Yup, I was just realizing that. We probably need to do the `-cover` dance for those too?

**Comment by ndeloof** [comment]
yes we probably need GOCOVERDIR set somehow

**Comment by EronWright** [comment]
Heads-up, I hit upon an issue related to `additional_contexts` and relative paths. 
https://github.com/docker/compose/issues/10448

Assumedly this is because the context is being passed uninterpreted by compose to buildx.

**Comment by ndeloof** [comment]
@EronWright according to the spec
https://github.com/compose-spec/compose-spec/blob/master/build.md

> Validation beyond that is the responsibility of the image builder

but indeed, it would make sense we detect and expand relative paths. I'll investigate if this is feasible

**Comment by TeKa-Dev** [comment]
I have a problem with `additional_contexts` and directory paths containing a dot at the beginning:
`
additional_contexts
  foo: /home/user/.bar/foo
`
but it works on buildx:
`docker buildx build . --build-context foo=/home/user/.bar/foo`

**Comment by laurazard** [comment]
Hiya @TeKa-Dev,

I can't seem to reproduce this:
```shell
$ ls /home/laurabrehm/.config/nvim/
init.lua  lazy-lock.json  LICENSE  lua  README.md  stylua.toml
```
```yaml
services:
  a:
    build:
        context: .
        additional_contexts:
          - dep=/home/laurabrehm/.config/nvim/
```
```Dockerfile
# syntax=docker/dockerfile:1
FROM alpine:latest
COPY --from=dep init.lua /
```
```shell
$ docker compose build
...
$ docker run --rm -it build-a sh
/ ls
...
init.lua
```

We might be able to help you if you provide a more complete reproduction scenario however.

**Comment by TeKa-Dev** [comment]
Hi, I was not accurate, not all directories with dot cause an error, but `.m2/repository/`

```bash
$ docker buildx build . --build-context m2repository=/home/teka/.m2/repository/
[+] Building 1.1s (12/12) FINISHED
 => [internal] load build definition from Dockerfile                                                 0.0s
 => => transferring dockerfile: 994B                                                                 0.0s
 => [internal] load .dockerignore                                                                    0.0s
 => => transferring context: 34B                                                                     0.0s
 => resolve image config for docker.io/docker/dockerfile:1.4                                         0.9s
 => CACHED docker-image://docker.io/docker/dockerfile:1.4@sha256:9ba7531bd80fb0a858632727cf7a112fbf  0.0s
 => [internal] load build definition from Dockerfile                                                 0.0s
 => [internal] load .dockerignore                                                                    0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                     0.0s
 => [context m2repository] load .dockerignore                                                        0.0s
 => => transferring m2repository: 102B                                                               0.0s
 => [context m2repository] load from client                                                          0.0s
 => => transferring m2repository: 406.09kB                                                           0.0s
 => CACHED [stage-0 1/2] FROM docker.io/library/alpine:latest                                        0.0s
 => [stage-0 2/2] COPY --from=m2repository org /                                                     0.0s
 => exporting to image                                                                               0.0s
 => => exporting layers                                                                              0.0s
 => => writing image sha256:c9a3f6a8e9a2f369d8cb5332c0f4652d44e35c52e622d2fc899851a3b30c87ac         0.0s
```

```yaml
services:
  alpine:
    build:
      context: .
      additional_contexts:
        - m2repository=/home/teka/.m2/repository/
```
```bash
$ docker-compose build
[+] Building 2.1s (11/11) FINISHED
 => [internal] load build definition from Dockerfile                                                 0.0s
 => => transferring dockerfile: 994B                                                                 0.0s
 => [internal] load .dockerignore                                                                    0.0s
 => => transferring context: 34B                                                                     0.0s
 => resolve image config for docker.io/docker/dockerfile:1.4                                         1.9s
 => CACHED docker-image://docker.io/docker/dockerfile:1.4@sha256:9ba7531bd80fb0a858632727cf7a112fbf  0.0s
 => [internal] load build definition from Dockerfile                                                 0.0s
 => [internal] load .dockerignore                                                                    0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                     0.0s
 => [context m2repository] load .dockerignore                                                        0.0s
 => => transferring m2repository: 2B                                                                 0.0s
 => [context m2repository] load from client                                                          0.0s
 => => transferring m2repository: 2B                                                                 0.0s
 => [stage-0 1/2] FROM docker.io/library/alpine:latest                                               0.0s
 => ERROR [stage-0 2/2] COPY --from=m2repository org /                                               0.0s
------
 > [stage-0 2/2] COPY --from=m2repository org /:
------
failed to solve: failed to compute cache key: "/org" not found: not found
```
I don't know if this is related but it doesn't work for me
```bash
$ docker compose build
services.alpine.build Additional property additional_contexts is not allowed
```
I use
```bash
$ docker -v
Docker version 20.10.23, build 7155243
$ docker-compose -v
Docker Compose version v2.17.2
```

**Pull request merged**